### ISSUE TITLE
Generics with 'extends' for return types cause type check failure.

### ISSUE THREAD

**Issue by na-o-ys** [Microsoft/TypeScript#13993]
**TypeScript Version:**  2.2.0

**Code**

```ts
interface I { attr: string }

type M = <T extends I>() => T
const v: M = () => ({})

v().attr
```

**Expected behavior:**

`Type '() => {}' is not assignable to type 'M'.
  Type '{}' is not assignable to type 'I'.`

**Actual behavior:**

The compilation is successful.

**Comment by aleksey-bykov** [comment]
@RyanCavanaugh is there any chance to fix assignability checks for generic and do it properly? not sure which issue tracks it (there are way to many)

**Comment by ahejlsberg** [comment]
Fixed by #16368.
